calico/felix/fetcd.py

```diff
@@ -60,6 +77,74 @@ def __init__(self, config):
         self.client = None
         self.my_config_dir = dir_for_per_host_config(self.config.HOSTNAME)
 
+        # Initialized at poll start time.
+        self.splitter = None
+        self.next_etcd_index = None
+
+        # Cache of known endpoints, used to resolve deletions of whole
+        # directory trees.
+        self.endpoint_ids_per_host = defaultdict(set)
+
+        self.handlers = {
+            "delete": self.force_resync,
+
+            "Ready": {
+                "set": self.on_ready,
+                "delete": self.force_resync,
+            },
+
+            "policy": {
+                "delete": self.force_resync,
+
+                "profile": {
+                    "delete": self.force_resync,
+
+                    "capture": ("profile_id",  {
+                        "delete": self.on_profile_delete,
+
+                        "tags": {
+                            "delete": self.on_tags_delete,
+                            "set": self.on_tags_set,
+                        },
+
+                        "rules": {
+                            "delete": self.on_rules_delete,
+                            "set": self.on_rules_set,
+                        }
+                    })
+                }
+            },
+
+            "host": {
+                "delete": self.force_resync,
+
+                "capture": ("hostname",  {
+                    "delete": self.on_host_delete,
+
+                    "workload": {
+                        "delete": self.on_host_delete,
+
+                        "capture": ("orchestrator",  {
+                            "delete": self.on_orch_delete,
+
+                            "capture": ("workload_id",  {
+                                "delete": self.on_workload_delete,
+
+                                "endpoint": {
+                                    "delete": self.on_workload_delete,
+
+                                    "capture": ("endpoint_id",  {
+                                        "delete": self.on_endpoint_delete,
+                                        "set": self.on_endpoint_set
+                                    })
+                                }
+                            })
+                        })
+                    }
+                })
+            }
+        }
+
```

Comment:
Oh, this is required by a customer.

:disappointed:

Fine, I suppose we have to merge this as-is. Doing this better really needs to be a top priority though.